components/script/dom/videotracklist.rs

```diff
@@ -0,0 +1,118 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::VideoTrackListBinding::{self, VideoTrackListMethods};
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::reflector::reflect_dom_object;
+use crate::dom::bindings::root::Dom;
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::videotrack::VideoTrack;
+use crate::dom::window::Window;
+use dom_struct::dom_struct;
+
+#[dom_struct]
+pub struct VideoTrackList {
+    eventtarget: EventTarget,
+    tracks: DomRefCell<Vec<Dom<VideoTrack>>>,
+}
+
+impl VideoTrackList {
+    pub fn new_inherited(tracks: &[&VideoTrack]) -> VideoTrackList {
+        VideoTrackList {
+            eventtarget: EventTarget::new_inherited(),
+            tracks: DomRefCell::new(tracks.iter().map(|track| Dom::from_ref(&**track)).collect()),
+        }
+    }
+
+    pub fn new(window: &Window, tracks: &[&VideoTrack]) -> DomRoot<VideoTrackList> {
+        reflect_dom_object(
+            Box::new(VideoTrackList::new_inherited(tracks)),
+            window,
+            VideoTrackListBinding::Wrap,
+        )
+    }
+
+    pub fn len(&self) -> usize {
+        self.tracks.borrow().len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.tracks.borrow().is_empty()
+    }
+
+    pub fn item(&self, idx: usize) -> Option<DomRoot<VideoTrack>> {
+        self.tracks
+            .borrow()
+            .get(idx)
+            .map(|track| DomRoot::from_ref(&**track))
+    }
+
+    pub fn selected_index(&self) -> Option<usize> {
+        self.tracks
+            .borrow()
+            .iter()
+            .position(|track| track.selected())
+    }
+
+    // TODO(#22799) Integrate DOM Audio and Video track selection with media player
+    pub fn set_selected(&self, idx: usize, value: bool) {
+        if let Some(track) = self.item(idx) {
+            track.set_selected(value);
+        }
+    }
+
+    pub fn add(&self, track: &VideoTrack) {
+        self.tracks.borrow_mut().push(Dom::from_ref(track));
```

Comment:
You may have mentioned this before but just to get some clarification, according to the spec it looks like VideoTrackList needs to somehow track the video tracks that were previously selected:

> Whenever a track in a VideoTrackList that was previously not selected is selected, and whenever the selected track in a VideoTrackList is unselected without a new track being selected in its stead, the user agent must queue a task to fire an event named change at the VideoTrackList object. This task must be queued before the task that fires the resize event, if any.

This seems like a field be added, such as a vector to track index's or a boolean field for a video track for being previously selected but I may be overthinking here.

Reply:
We can either add a new index member that stores the currently selected index, or we can scan the list of tracks each time.